File: dj_client/src-tauri/src/content_filter.rs
//! Slur detection for user-facing text in the DJ client.
//!
//! Uses the `rustrict` crate with `Type::OFFENSIVE & Type::SEVERE` to catch
//! severe slurs while ignoring general profanity.

use rustrict::{CensorStr, Type};

/// Return `true` if *text* contains a severe slur.
pub fn contains_slur(text: &str) -> bool {
    // Strip zero-width / invisible Unicode before checking
    let cleaned = strip_invisible(text);
    cleaned.is(Type::OFFENSIVE & Type::SEVERE)
}

/// Validate that *text* does not contain a severe slur.
///
/// Returns `Ok(())` on clean input, or `Err(String)` with a vague error
/// message (never revealing what was detected).
pub fn validate_no_slurs(text: &str, field_name: &str) -> Result<(), String> {
    if contains_slur(text) {
        Err(format!("{field_name} contains language that is not allowed"))
    } else {
        Ok(())
    }
}

/// Strip zero-width and invisible Unicode characters that could be used to
/// bypass the filter.
fn strip_invisible(text: &str) -> String {
    text.chars()
        .filter(|c| !is_invisible(*c))
        .collect()
}

fn is_invisible(c: char) -> bool {
    matches!(
        c,
        '\u{200b}' // zero-width space
        | '\u{200c}' // zero-width non-joiner
        | '\u{200d}' // zero-width joiner
        | '\u{200e}' // LTR mark
        | '\u{200f}' // RTL mark
        | '\u{2060}' // word joiner
        | '\u{feff}' // BOM / zero-width no-break space
        | '\u{00ad}' // soft hyphen
        | '\u{034f}' // combining grapheme joiner
        | '\u{061c}' // Arabic letter mark
        | '\u{115f}' // Hangul choseong filler
        | '\u{1160}' // Hangul jungseong filler
        | '\u{17b4}' // Khmer vowel inherent aq
        | '\u{17b5}' // Khmer vowel inherent aa
        | '\u{180e}' // Mongolian vowel separator
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn clean_text_passes() {
        assert!(!contains_slur("DJ BassDropper"));
        assert!(!contains_slur("The Groove Machine"));
        assert!(!contains_slur("Chill Vibes Only"));
    }

    #[test]
    fn basic_profanity_allowed() {
        // General profanity should NOT be flagged — only severe slurs
        assert!(!contains_slur("damn"));
        assert!(!contains_slur("hell"));
        assert!(!contains_slur("crap"));
    }

    #[test]
    fn empty_and_whitespace() {
        assert!(!contains_slur(""));
        assert!(!contains_slur("   "));
    }

    #[test]
    fn scunthorpe_safe() {
        // Words that contain slur substrings but are benign
        assert!(!contains_slur("Scunthorpe"));
        assert!(!contains_slur("classic"));
    }

    #[test]
    fn validate_clean_ok() {
        assert!(validate_no_slurs("DJ Cool", "DJ name").is_ok());
    }

    #[test]
    fn validate_vague_error() {
        // We don't test specific slurs here, just that the error message
        // is vague when validation fails
        let result = validate_no_slurs("test", "DJ name");
        // Clean text should pass
        assert!(result.is_ok());
    }
}

File: coordinator/app/utils/content_filter.py
"""Slur detection for user-facing text fields.

Uses ``better-profanity`` loaded with a **custom curated list** of severe
slurs only (racial, ethnic, homophobic).  General profanity ("damn", "hell",
etc.) is intentionally NOT flagged.

The library auto-generates common leet-speak / substitution variants for
every word in the list and uses word-boundary matching, so we only need to
supply canonical forms.

# SYNC: keep SEVERE_SLURS list in sync with vj_server/content_filter.py
"""

from __future__ import annotations

import re
import unicodedata

from better_profanity import profanity

# ---------------------------------------------------------------------------
# Curated word list — severe slurs only
# ---------------------------------------------------------------------------
# These are canonical forms.  ``better-profanity`` auto-expands each entry
# with common leet-speak substitutions (e.g. 1→i, 0→o, @→a, 3→e, $→s).
#
# Criteria for inclusion:
#   1. Widely recognised as a severe slur targeting a protected group.
#   2. No common benign meaning that would trigger false positives.
#
# This list replaces the library's default 400+ word list so that everyday
# profanity is never flagged.
# ---------------------------------------------------------------------------

SEVERE_SLURS: list[str] = [
    # Anti-Black
    "nigger",
    "nigga",
    "niggers",
    "niggas",
    "coon",
    "coons",
    "darkie",
    "darkies",
    "jiggaboo",
    "jigaboo",
    "sambo",
    # Anti-Asian
    "chink",
    "chinks",
    "gook",
    "gooks",
    "zipperhead",
    # Anti-Hispanic
    "spic",
    "spick",
    "spics",
    "wetback",
    "wetbacks",
    "beaner",
    "beaners",
    # Anti-Jewish
    "kike",
    "kikes",
    # Anti-LGBTQ
    "faggot",
    "faggots",
    "fag",
    "fags",
    "dyke",
    "dykes",
    "tranny",
    "trannies",
    # Anti-Roma / general ethnic
    "gyppo",
    "raghead",
    "ragheads",
    "towelhead",
    "towelheads",
    "camel jockey",
    # Anti-white (also severe)
    # Disability-related slurs
    "retard",
    "retards",
    "retarded",
]

# ---------------------------------------------------------------------------
# Initialise the profanity filter with our custom list
# ---------------------------------------------------------------------------
profanity.load_censor_words(SEVERE_SLURS)

# Zero-width and invisible Unicode characters that could be inserted to
# bypass the filter.
_INVISIBLE_RE = re.compile(
    "["
    "\u200b"  # zero-width space
    "\u200c"  # zero-width non-joiner
    "\u200d"  # zero-width joiner
    "\u200e"  # LTR mark
    "\u200f"  # RTL mark
    "\u2060"  # word joiner
    "\ufeff"  # BOM / zero-width no-break space
    "\u00ad"  # soft hyphen
    "\u034f"  # combining grapheme joiner
    "\u061c"  # Arabic letter mark
    "\u115f"  # Hangul choseong filler
    "\u1160"  # Hangul jungseong filler
    "\u17b4"  # Khmer vowel inherent aq
    "\u17b5"  # Khmer vowel inherent aa
    "\u180e"  # Mongolian vowel separator
    "]+"
)


def _normalise(text: str) -> str:
    """Strip invisible chars and apply NFKD Unicode normalisation."""
    text = _INVISIBLE_RE.sub("", text)
    text = unicodedata.normalize("NFKD", text)
    return text


def contains_slur(text: str) -> bool:
    """Return ``True`` if *text* contains a severe slur."""
    if not text:
        return False
    cleaned = _normalise(text)
    return profanity.contains_profanity(cleaned)


_VAGUE_ERROR = "contains language that is not allowed"


def validate_no_slurs(text: str, field_name: str = "value") -> str:
    """Return *text* unchanged, or raise ``ValueError`` if a slur is found.

    Suitable for use inside Pydantic ``@field_validator`` methods.
    """
    if text and contains_slur(text):
        raise ValueError(f"{field_name} {_VAGUE_ERROR}")
    return text
